hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSClient.java

```diff
@@ -922,14 +922,21 @@ boolean renewLease() throws IOException {
         updateLastLeaseRenewal();
         return true;
       } catch (IOException e) {
-        // Abort if the lease has already expired. 
+        // Abort if the lease has already expired.
         final long elapsed = Time.now() - getLastLeaseRenewal();
         if (elapsed > HdfsConstants.LEASE_HARDLIMIT_PERIOD) {
           LOG.warn("Failed to renew lease for " + clientName + " for "
               + (elapsed/1000) + " seconds (>= hard-limit ="
               + (HdfsConstants.LEASE_HARDLIMIT_PERIOD/1000) + " seconds.) "
               + "Closing all files being written ...", e);
           closeAllFilesBeingWritten(true);
+        }
+        // Abort if token has already expired
+        else if (e.getMessage() != null && e.getMessage()
+                .matches("token \\(HDFS_DELEGATION_TOKEN token.*\\) (can't be found in cache|is expired)")) {
```

Comment:
It would be better to include this matching logic in the type of Exception returned.